### PR TITLE
[TwigBundle]  Pre-compile only *.twig files in cache warmup

### PR DESCRIPTION
**Pull request by GromNaN**
| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

In my project, the template directory contains all the frontend assets (CSS, JS and images), this is a choice to locate all component files in the same directories. But the asset files are compiled as if they were twig template, which is a waste of time and storage.

```
front/
├─ components/
│  ├─ video/
│  │  ├─ templates/
│  │  │  ├─ video.html.twig
│  │  ├─ styles/
│  │  │  ├─ video.css
│  │  ├─ scripts/
│  │  │  ├─ video.js
```

~This patch limit warmup to `*.twig` files.~
This PR adds an option to restrict file pattern, for cache warmup and lint.

```yaml
twig:
    file_name_pattern: "*.twig"
```

|                  | Before | After |
|------------------|--------|-------|
| Time to generate | 25s    | 9s    |
| Storage          | 80MB   | 26MB  |
| Number of files  | 4800   | 1500  |

The filter on `*.twig` files is already in the `lint:twig` command.

Note: If a non-twig template is included in a template (ex: to inline CSS), the `source()` function must be used ; the cache is not used in this case.

Edit: Here is the documentation https://symfony.com/doc/current/reference/configuration/twig.html#file-name-pattern

### CODE REVIEW

**Comment by fabpot** [comment]
I'm not sure about this change, but this can only be merged in 6.1 anyway as this is a behavior change and not a bug fix (we always consider performance improvements as a new feature).

**Comment by GromNaN** [comment]
Ok for 6.1.

The question to answer before merging this PR is: is it expected to have twig templates that does not have the `.twig` extension?

---

The alternative in my code is to decorate the iterator and filter each template name.
```php
class TemplateIterator implements \IteratorAggregate
{
    public function __construct(private iterable $iterator) {}
    public function getIterator(): \Generator
    {
        foreach ($this->iterator as $filename) {
            if (str_ends_with($filename, '.twig')) {
                yield $filename;
            }
        }
    }
}
```

And use a compiler pass to inject the decorator in the cache warmer.
```php
class TemplateIteratorPass implements CompilerPassInterface
{
    public function process(ContainerBuilder $container)
    {
        $warmer = $container->getDefinition('twig.template_cache_warmer');
        $iterator = $container->register(TemplateIterator::class, TemplateIterator::class)->addArgument($warmer->getArgument(1));
        $warmer->setArgument(1, $iterator);
    }
}
```

**Comment by fabpot** [comment]
> Ok for 6.1.
> 
> The question to answer before merging this PR is: is it expected to have twig templates that does not have the `.twig` extension?
> 

I think that this is indeed a great question. And the answer is: you can have templates with another extension. It's probably not widespread but nothing forbids it.

**Comment by GromNaN** [comment]
I added an option to the `TwigBundle`, ~the plan is to change its default value in 7.0~:

```yaml
twig:
    file_name_pattern: "*.twig"
```

~The majority of projects having only `*.twig` files in the template directories, they will not have the deprecation notice and not need to update their configuration.~

~For projects that mix file extensions (like mine), the deprecation notice will be triggered and it will be easy to solve by adding the config.~

**Comment by GromNaN** [comment]
Unfortunately, the deprecation is not reported in logs when running `bin/console cache:warmup`. And there is very few chances the cache is warmed in tests. I've removed the deprecation to simply add the option. Updating the default value in 7.0 could be part of an other PR.

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -141,6 +141,15 @@ private function addTwigOptions(ArrayNodeDefinition $rootNode)
                     ->info('The default path used to load templates')
                     ->defaultValue('%kernel.project_dir%/templates')
                 ->end()
+                ->arrayNode('file_name_pattern')
+                    ->example('*.twig')
+                    ->info('Pattern of file name used for cache warmer')
```
Comment (GromNaN) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/Configuration.php]: This option should be used for linter too.

**Comment by GromNaN** [comment]
> > Ok for 6.1.
> > The question to answer before merging this PR is: is it expected to have twig templates that does not have the `.twig` extension?
> 
> I think that this is indeed a great question. And the answer is: you can have templates with another extension. It's probably not widespread but nothing forbids it.

In Symfony WebProfilerBundle itself, all svg are included as templates with `include` tag.
https://github.com/symfony/symfony/blob/24304ad77874f03f9971fd9e29c0ef0776b0f5b9/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/header.html.twig#L3

https://github.com/symfony/symfony/blob/24304ad77874f03f9971fd9e29c0ef0776b0f5b9/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/time.html.twig#L150

Zurb templates uses the `source()` function correctly.
https://github.com/symfony/symfony/blob/24304ad77874f03f9971fd9e29c0ef0776b0f5b9/src/Symfony/Bridge/Twig/Resources/views/Email/zurb_2/notification/body.html.twig#L6-L7

See https://github.com/symfony/symfony/pull/45858

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -141,6 +141,15 @@ private function addTwigOptions(ArrayNodeDefinition $rootNode)
                     ->info('The default path used to load templates')
                     ->defaultValue('%kernel.project_dir%/templates')
                 ->end()
+                ->arrayNode('file_name_pattern')
+                    ->example('*.twig')
+                    ->info('Pattern of file name used for cache warmer and linter')
+                    ->beforeNormalization()
+                        ->ifString()
+                            ->then(function ($value) { return [$value]; })
```
Comment (GromNaN) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/Configuration.php:149]: Arrow function.

File: src/Symfony/Bridge/Twig/Command/LintCommand.php [hunk 3]
```diff
@@ -41,12 +41,14 @@ class LintCommand extends Command
 {
     private Environment $twig;
     private string $format;
+    protected array $namePatterns;
 
-    public function __construct(Environment $twig)
+    public function __construct(Environment $twig, array $namePatterns = ['*.twig'])
```
Comment (GromNaN) [inline src/Symfony/Bridge/Twig/Command/LintCommand.php]: Could be `string|array`

File: src/Symfony/Bundle/TwigBundle/TemplateIterator.php [hunk 4]
```diff
@@ -29,16 +29,19 @@ class TemplateIterator implements \IteratorAggregate
...
     private array $paths;
     private ?string $defaultPath;
+    private array $fileNames;
 
     /**
-     * @param array       $paths       Additional Twig paths to warm
-     * @param string|null $defaultPath The directory where global templates can be stored
+     * @param array         $paths       Additional Twig paths to warm
+     * @param string|null   $defaultPath The directory where global templates can be stored
+     * @param string[]|null $fileNames   Pattern of file names
```
Comment (GromNaN) [inline src/Symfony/Bundle/TwigBundle/TemplateIterator.php]: Not nullable. 
Empty array = no filter.

Comment (fabpot) [reply src/Symfony/Bundle/TwigBundle/TemplateIterator.php]: So, null should be removed here?

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php [hunk 5]
```diff
@@ -97,6 +97,9 @@ public function load(array $configs, ContainerBuilder $container)
         // paths are modified in ExtensionPass if forms are enabled
         $container->getDefinition('twig.template_iterator')->replaceArgument(1, $config['paths']);
 
+        $container->getDefinition('twig.template_iterator')->replaceArgument(3, $config['file_name_pattern']);
+        $container->getDefinition('twig.command.lint')->replaceArgument(1, $config['file_name_pattern'] ?: ['*.twig']);
```
Comment (fabpot) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php:101]: Should the default be "*" for BC? If we keep "*.twig" as a default, this should at least be documented in the CHANGELOG file.

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Bridge/Twig/Command/LintCommand.php [hunk 6]
```diff
@@ -41,12 +41,14 @@ class LintCommand extends Command
 {
     private Environment $twig;
     private string $format;
+    protected array $namePatterns;
```
Comment (stloyd) [inline src/Symfony/Bridge/Twig/Command/LintCommand.php]: Shouldn't protected be before private same as with methods?

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php (hunk 5 above)
Comment (GromNaN) [reply src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php:101]: For the linter, I kept `*.twig` for BC.

**Comment by yceruto** [comment]
For the record and to not miss something in previous discussions:
 * https://github.com/symfony/symfony/issues/28282
 * https://github.com/symfony/symfony/pull/34321
 * https://github.com/symfony/symfony/issues/34142
 * https://github.com/symfony/symfony/pull/34416
 * https://github.com/symfony/symfony/issues/41950

**Comment by GromNaN** [comment]
Thanks @yceruto. This proves this issue is common. Your PR was almost the same, with excluded directories in addition.

@fabpot [presented the argument](https://github.com/symfony/symfony/pull/34416#issuecomment-554723176) that `templates` directory is for templates only, but experience shows that grouping files by their type is not optimal for maintenance. A twig loader namespace can target a directory with several `templates` directories.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @GromNaN.

**Pull request merged**